### PR TITLE
gh-123083: Fix a potential use-after-free in ``STORE_ATTR_WITH_HINT``

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-123083

### CODE REVIEW

File: Lib/test/test_dict.py [hunk 1]
```diff
@@ -1476,6 +1476,24 @@ def test_dict_items_result_gc_reversed(self):
         gc.collect()
         self.assertTrue(gc.is_tracked(next(it)))
 
+    def test_store_evilattr(self):
```
Comment (corona10) [inline Lib/test/test_dict.py:1479]: I am not sure that this is the proper place.

File: Misc/NEWS.d/next/Core_and_Builtins/2024-08-17-17-26-25.gh-issue-123083.9xWLJ-.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix STORE_ATTR_WITH_HINT has potential use-after-free.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-08-17-17-26-25.gh-issue-123083.9xWLJ-.rst]:
```suggestion
Fix a potential use-after-free in ``STORE_ATTR_WITH_HINT``.
```

Comment (corona10) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-08-17-17-26-25.gh-issue-123083.9xWLJ-.rst]: Thanks, I will address the changes after build bot tests are passed.

File: Python/bytecodes.c [hunk 3]
```diff
@@ -2234,12 +2234,12 @@ dummy_func(
             PyDict_WatchEvent event = old_value == NULL ? PyDict_EVENT_ADDED : PyDict_EVENT_MODIFIED;
             new_version = _PyDict_NotifyEvent(tstate->interp, event, dict, name, PyStackRef_AsPyObjectBorrow(value));
```
Comment (colesbury) [inline Python/bytecodes.c]: I think we should structure this to match the ordering in `dictobject.c`:

https://github.com/python/cpython/blob/b6d0a40a42fa17d087e502245b29cde46368ac07/Objects/dictobject.c#L1693-L1707

1) Update GC tracking if necessary
2) Notify event
3) Store value and version tag
4) Decref old value

**Comment by markshannon** [comment]
Could you add a comment, specifically why this is the correct order?
It isn't obvious and it will be easy to reintroduce the error without some warning not to.

**Comment by corona10** [comment]
@colesbury @markshannon gentle ping :)

**Review by markshannon — APPROVED** [review]
This looks fine to me. But let @colesbury review it before merging.

**Review by colesbury — APPROVED** [review]
LGTM too

**Pull request merged**